Feature: Add NotRoadTypes (NRT)

andythenorth opened the pull request:
The 'one big diff' approach.

16 types only, PeterN has a patch for 64, but that needs finish, tested, argued about etc.  Can do that later.  We make progress by not waiting for perfect eh? :)  

Spec: https://wiki.openttd.org/Frosch/NotRoadTypes

Want to try it out? https://www.openttd.org/downloads/openttd-pullrequests/pr6811/latest.html has precompiled binaries for you! (last updated: 24th of March 2018)

PeterN committed:
Codechange: Untangle game and scenario toolbars.

PeterN commented:
Argument for doing 64 types initially is it saves having to add a bunch of savegame conversion code.

andythenorth commented:
Previous reviews by PeterN and TrueBrain took place in a ticket of the NRT fork (predated OpenTTD github).

https://github.com/andythenorth/NotRoadTypes/issues/22#issuecomment-379849291

andythenorth commented:
Rebased by Alberth using Peter's July 2018 rebase
https://github.com/Alberth289346/OpenTTD/tree/nrt-block-based

This builds and runs fine, but it's not clear yet:

- if there are any compile warnings to attend to
- whether docs (landscape_grid.html etc) are accurate
- what state translation strings are in

With respect to 64 types (or not): this PR adds 64 types.  It's the better option.

Crowd-sourced opinion is that this won't have enough play-test time to make it into 1.9.0 by April 2019.  It has been play-tested quite a lot already, but eh.  Save it for 2.0? 🎁

nielsmh reviewed (dismissed):
Assertion failure at tile_map.h line 91 running regression tests.

https://github.com/andythenorth/OpenTTD/blob/f5b1115039e41d7f75f91e7355003f10dd01b352/src/tile_map.h#L89-L93

PeterN commented:
I have rebased to master, available here https://github.com/PeterN/OpenTTD/tree/nrt-block

New sprites have been split off to separate files to avoid future conflicts. I have not addressed the regression test failure.

nielsmh commented:
The assertion failure mentioned above still happens in the regression tests.

PeterN commented:
> 
> 
> The assertion failure mentioned above still happens in the regression tests.

Yes, #7142 has only just been merged into master!

PeterN commented:
Two issues I've noticed during gameplay testing:

* Inconsistencies between roadtypes available and roadtypes buildable, resulting in misleading menus and button states. This may be something to do with date-introduced roadtypes.
* Not possible to set up one-way roads.

PeterN commented:
One-way roads have now been fixed, this was due to the 64 road types addition. The regression failure is now resolved!

TrueBrain commented:
The latest version at the time of writing can be found in binary form here:
https://www.openttd.org/downloads/openttd-pullrequests/pr6811/latest.html

I will try to keep it up-to-date where possible

PeterN commented:
Need to fix the AI changelog as it's all under 1.9 still :-)

Gymnasiast commented on src/road.h:
```diff
@@ -0,0 +1,279 @@
+	ROTF_CATENARY = 0,                                     ///< Bit number for adding catenary
+	ROTF_NO_LEVEL_CROSSING,                                ///< Bit number for disabling level crossing
+	ROTF_NO_HOUSES,                                        ///< Bit number for setting this roadtype as not house friendly
+	ROTF_HIDDEN,                                           ///< Bit number for hiden from construction.
```
Typo in "hiden".
```suggestion
	ROTF_HIDDEN,                                           ///< Bit number for hidden from construction.
```

andythenorth commented:
Merge it then :)

LordAro commented on bin/ai/regression/tst_regression/main.nut:
```diff
@@ -535,14 +535,14 @@ function Regression::Prices()
-	print("  ROADTYPE_TRAM,BT_DEPOT:      " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_DEPOT));
-	print("  ROADTYPE_TRAM,BT_BUS_STOP:   " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_BUS_STOP));
-	print("  ROADTYPE_TRAM,BT_TRUCK_STOP: " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_TRUCK_STOP));
+	print("  ROADTYPE_ROAD,0,BT_ROAD:       " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_ROAD, 0, AIRoad.BT_ROAD));
```
spacing

PeterN replied:
This is autogenerated code

LordAro replied:
...no it isn't? from where?

PeterN replied:
Oh you're right, just confused me by being in bin :-)

PeterN replied:
Oh wait, you are referring to the ,0, part. That's based on the existing spacing here and above for -Rail-.

LordAro commented on src/script/api/script_road.hpp:
```diff
@@ -53,6 +58,14 @@ class ScriptRoad : public ScriptObject {
+	/**
+	 * Subtypes of road known to the game.
+	 */
+	enum RoadSubType {
```
All roadtypes have at least one subtype (0) right? Perhaps adding a constant for that would be nicer than writing "0" for the subtype for all the script functions? (I'm looking at the regression tests while writing this)

PeterN replied:
This was modeled on the railtypes API which also has no default values.

PeterN replied:
RoadSubTypes are gone now!

LordAro commented on src/autoreplace_gui.cpp:
```diff
@@ -411,9 +426,19 @@ class ReplaceVehicleWindow : public Window {
+				break;
+
+			default:
+				break;
```
not consistent with others, should be NOT_REACHED or not here at all

LordAro commented on src/engine_type.h:
```diff
@@ -123,6 +125,7 @@ struct RoadVehicleInfo {
 	uint8 air_drag;          ///< Coefficient of air drag
 	byte visual_effect;      ///< Bitstuffed NewGRF visual effect data
 	byte shorten_factor;     ///< length on main map for this type is 8 - shorten_factor
+	RoadSubType roadsubtype;
```
doxyment

LordAro commented on src/ground_vehicle.cpp:
```diff
@@ -28,7 +28,7 @@ void GroundVehicle<T, Type>::PowerChanged()
 	uint32 max_te = 0;
 	uint32 number_of_parts = 0;
-	uint16 max_track_speed = v->GetDisplayMaxSpeed();
+	uint16 max_track_speed = this->vcache.cached_max_speed; // Note: Different units for RV and Trains.
```
why this change? comment doesn't really explain why...

LordAro commented on src/newgrf.cpp:
```diff
@@ -5482,6 +5745,36 @@ static void RailTypeMapSpriteGroup(ByteReader *buf, uint8 idcount)
 
+static void RoadTypeMapSpriteGroup(ByteReader *buf, uint8 idcount, RoadType basetype)
+{
+	uint8 *roadtypes = AllocaM(uint8, idcount);
```
alloca make me sad :(
but fixing is for another PR, i think

LordAro commented on src/newgrf.cpp:
```diff
@@ -5482,6 +5745,36 @@ static void RailTypeMapSpriteGroup(ByteReader *buf, uint8 idcount)
+{
+	uint8 *roadtypes = AllocaM(uint8, idcount);
+	for (uint i = 0; i < idcount; i++) {
+		roadtypes[i] = _cur.grffile->roadtype_map[basetype][buf->ReadByte()];
```
no bounds check?

LordAro commented on src/newgrf.cpp:
```diff
@@ -5884,13 +6185,20 @@ static void GraphicsNew(ByteReader *buf)
 	/* Load <num> sprites starting from <replace>, then skip <skip_num> sprites. */
 	grfmsg(2, "GraphicsNew: Replacing sprites %d to %d of %s (type 0x%02X) at SpriteID 0x%04X", offset, offset + num - 1, action5_type->name, type, replace);
 
+	if (type == 0x0D) _loaded_newgrf_features.shore = SHORE_REPLACE_ACTION_5;
```
why was this moved?

LordAro commented on src/newgrf.cpp:
```diff
@@ -9172,6 +9497,21 @@ static void ActivateOldShore()
 }
 
+/**
+ * Replocate the old tram depot sprites to the new position, if no new ones were loaded.
```
replocate

LordAro commented on src/newgrf_roadtype.cpp:
```diff
@@ -0,0 +1,139 @@
+RoadTypeResolverObject::RoadTypeResolverObject(const RoadtypeInfo *rti, TileIndex tile, TileContext context, RoadTypeSpriteGroup rtsg, uint32 param1, uint32 param2)
+	: ResolverObject(rti != NULL ? rti->grffile[rtsg] : NULL, CBID_NO_CALLBACK, param1, param2), roadtype_scope(*this, tile, context)
+{
+	this->root_spritegroup = rti != NULL ? rti->group[rtsg] : NULL;
```
why not put this in the initialiser list above?

PeterN replied:
Well, I can't get that working...
`error: member initializer 'root_spritegroup' does not name a non-static data member or base class`

LordAro commented on src/pathfinder/npf/npf.cpp:
```diff
@@ -860,7 +872,7 @@ static void NPFFollowTrack(AyStar *aystar, OpenListNode *current)
 	/* Information about the vehicle: TransportType (road/rail/water) and SubType (compatible rail/road types) */
 	TransportType type = user->type;
-	uint subtype = user->roadtypes;
+	uint subtype = user->roadtype;
```
this doesn't appear to be a "subtype" - variable needs renaming?

LordAro commented on src/road_cmd.cpp:
```diff
@@ -31,15 +31,151 @@
+	static const RoadtypeInfo empty_roadtype = {
+		{0,0,0,0,0,0},
+		{0,0,0,0,0,0},
+		{0,0,0,0,0,0,0,0,0,{},{},0,{},{}},
```
spacing

LordAro commented on src/road_cmd.cpp:
```diff
@@ -1174,6 +1359,33 @@ const byte _road_sloped_sprites[14] = {
+{
+	if (slope != SLOPE_FLAT) {
+		switch (slope) {
+			case SLOPE_NE: return 11;
```
magic numbers?

LordAro commented on src/road_gui.cpp:
```diff
@@ -309,15 +266,20 @@ static bool RoadToolbar_CtrlChanged(Window *w)
-	int last_started_action; ///< Last started user action.
+	RoadTypeIdentifier roadtype_identifier; ///< Road type to build.
+	const RoadtypeInfo *rti;          ///< Informations about current road type
+	int last_started_action;    ///< Last started user action.
```
alignment

LordAro commented on src/road_gui.cpp:
```diff
@@ -713,15 +758,32 @@ static Hotkey roadtoolbar_hotkeys[] = {
-HotkeyList BuildRoadToolbarWindow::hotkeys("roadtoolbar", roadtoolbar_hotkeys, RoadToolbarGlobalHotkeys);
+HotkeyList BuildRoadToolbarWindow::road_hotkeys("roadtoolbar", roadtoolbar_hotkeys, RoadToolbarGlobalHotkeys);
+
+static Hotkey tramtoolbar_hotkeys[] = {
```
isn't this just identical to roadtoolbar_hotkeys? Can it be reused instead of duplicating?

LordAro commented on src/road_gui.cpp:
```diff
@@ -847,17 +916,54 @@ static WindowDesc _build_road_scen_desc(
+	EndContainer(),
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_IMGBTN, COLOUR_DARK_GREEN, WID_ROT_ROAD_X),
+						SetFill(0, 1), SetMinimalSize(22, 22), SetDataTip(SPR_IMG_TRAMWAY_X_DIR, STR_ROAD_TOOLBAR_TOOLTIP_BUILD_TRAMWAY_SECTION),
```
4 tabs seems like a lot.. is that consistent with what's done elsewhere?

LordAro commented on src/road_gui.cpp:
```diff
@@ -1146,11 +1253,93 @@ static WindowDesc _tram_station_picker_desc(
 }
+
+/**
+ * I really don't know why rail_gui.cpp has this too, shouldn't be included in the other one?
```
uh...

LordAro commented on src/saveload/afterload.cpp:
```diff
@@ -1045,18 +1061,18 @@ bool AfterLoadGame()
 						case ROAD_TILE_DEPOT:    break;
 					}
-					SetRoadTypes(t, ROADTYPES_ROAD);
+					SB(_me[t].m7, 6, 2, ROADTYPES_ROAD);
```
helper function for this?

PeterN replied:
No, this is old savegame conversion code. At this point it is converting a pre-tram game to post-tram format, then later on it will convert this pre-nrt format to post-nrt format.

LordAro commented on src/script/api/script_road.cpp:
```diff
@@ -520,11 +565,11 @@ static bool NeighbourHasReachableRoad(::RoadTypes rts, TileIndex start_tile, Dia
 	}
 
-	uint p2 = station_id == ScriptStation::STATION_JOIN_ADJACENT ? 0 : 32;
+	uint p2 = (station_id != ScriptStation::STATION_JOIN_ADJACENT) << 2;
```
think i'd still prefer a ternary for this

LordAro commented on src/script/api/script_road.hpp:
```diff
@@ -71,6 +84,19 @@ class ScriptRoad : public ScriptObject {
+	 * @note Since there is no string with only the name of the road, the text which
+	 *  is shown in the dropdown where you can chose a road type is returned. This
+	 *  means that the name could be something like "Road construction" instead
+	 *  of just "Road".
```
this seems like a deficiency... can this not be solved?

PeterN replied:
Yeah, seems odd to not include it

PeterN replied:
I think it's wrong, too, there is a string with just the name.

PeterN replied:
Yeah, it was added later. I wonder if it's mandatory. Hmm.

LordAro commented on src/smallmap_gui.cpp:
```diff
@@ -478,6 +478,23 @@ static inline uint32 GetSmallMapRoutesPixels(TileIndex tile, TileType t)
 
 		const SmallMapColourScheme *cs = &_heightmap_schemes[_settings_client.gui.smallmap_land_colour];
 		return ApplyMask(cs->default_colour, &andor);
+	} else if (t == MP_ROAD) {
```
should be transformed into a switch?

PeterN replied:
Possibly. It does create a quite a bit of extra identation.

LordAro commented on src/table/roadtypes.h:
```diff
@@ -0,0 +1,185 @@
+/**
+ * Global Roadtype definition
+ */
+static const RoadtypeInfo _original_roadtypes[] = {
```
do these particularly need to be arrays? given there's one, and will only ever be one

PeterN replied:
Good point. There were 2 at one point but I took the decision to undo that, based on our content-via-newgrf policy.

LordAro commented on src/town_cmd.cpp:
```diff
@@ -1174,6 +1206,41 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
+	const TileIndexDiffC *ptr;
+	bool allow = false;
+
+	for (ptr = tiles; ptr != endof(tiles); ++ptr) {
```
ptr should be declared here?

LordAro reviewed (commented):
So this is a beast of a PR. It's really hard to review in its current state, and personally I'd like to see the "Feature" commit split up further - there's some stuff about transparency, some other stuff i've forgotten because i've been reading this code for 30 mins now, and just generally lots of moving code about

Generally some doxygen comments are missing, would be good to get some of those filled out

PeterN commented:
Only thing I see about transparency is to do with hiding catenary, which is done the same way as for elrail.

I would have liked it to be split up as well, however we had already wasted lots of time trying to disentangle the original branch, and then indeed find parts of that could be neatly separated.

PeterN committed:
Cleanup: Update modified and removed strings in language files.

PeterN committed:
Feature: Add NotRoadTypes (NRT)

PeterN committed:
Add: New RoadType API functions.

PeterN committed:
Codechange: MakeBridgeRamp()'s road-or-rail-type parameter does not make sense.

Road type and rail type are stored in separate locations, so this parameter does
not make make sense as it is only used for rail bridges. Instead explicitly set the
rail type in MakeRailBridgeRamp().

PeterN committed:
Codechange: Use RAILTYPES_NONE instead of INVALID_RAILTYPES when not using rail.

INVALID_RAILTYPES, if it was accidentally tested, would match any railtype.

PeterN commented:
I guess I need to come clean on this, as it seems nobody knows: this is awaiting review and approval for merging.

stormcone commented:
If you try to "autoreplace" a ship or an aircraft, the game crashes:
"NOT_REACHED triggered at line 460 of OpenTTD-nrt-block/src/autoreplace_gui.cpp"

PeterN commented:
@stormcone Thanks, this has now been fixed.

Gymnasiast commented:
Congratulations on getting this massive update finished! :tada:
